cpu/fe310/periph/spi.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (C) 2019 Tristan Bruns
+ *               2019 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_fe310
+ * @ingroup     drivers_periph_spi
+ *
+ * @{
+ *
+ * @file        spi.c
+ * @brief       Low-level SPI driver implementation
+ *
+ * @author      Tristan Bruns
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "mutex.h"
+#include "assert.h"
+#include "periph/spi.h"
+
+#include "vendor/spi.h"
+
+#define ENABLE_DEBUG        (0)
+#include "debug.h"
+
+static const uint32_t _spi_clks[] = {
+    100000,
+    400000,
+    1000000,
+    5000000,
+    10000000,
+};
+
+#define SPI_CLK_NUMOF       ARRAY_SIZE(_spi_clks)
+
+static uint32_t _spi_clks_config[SPI_CLK_NUMOF] = { 0 };
+
+/* DIV_UP is division which rounds up instead of down */
+#define SPI_DIV_UP(a,b) (((a) + ((b) - 1)) / (b))
+
+/**
+ * @brief   Allocation device locks
+ */
+static mutex_t lock;
+
+void spi_init(spi_t dev)
+{
+    /* make sure given bus device is valid */
+    assert(dev < SPI_NUMOF);
+
+    /* initialize the buses lock */
+    mutex_init(&lock);
+
+    for (uint8_t i = 0; i < SPI_CLK_NUMOF; ++i) {
+        _spi_clks_config[i] = SPI_DIV_UP(cpu_freq(), 2 * _spi_clks[i]) - 1;
+    }
+
+    /* trigger pin initialization */
+    spi_init_pins(dev);
+
+    /* disable hardware chip select
+       (hardware chip select only supports one-byte transfers...) */
+    _REG32(spi_config[dev].addr, SPI_REG_CSMODE) = SPI_CSMODE_OFF;
+}
+
+void spi_init_pins(spi_t dev)
+{
+    assert(dev < SPI_NUMOF);
+
+    const gpio_t spi1_pins =
+        (1 << spi_config[dev].mosi) |
+        (1 << spi_config[dev].miso) |
+        (1 << spi_config[dev].sclk);
+
+    /* Enable I/O Function 0 */
+    GPIO_REG(GPIO_IOF_EN)  |=  spi1_pins;
+    GPIO_REG(GPIO_IOF_SEL) &= ~spi1_pins;
+}
+
+int spi_init_cs(spi_t dev, spi_cs_t cs)
+{
+    (void)dev;
+    assert(dev < SPI_NUMOF);
+
+    /* setting the CS high before configuring it as an
+       output should be fine on FE310. */
```

Comment:
Can you explain this?

Reply:
hmm, not really, this was already implemented this way in the original PR. Maybe @pyropeter can give some insight ?